// @ndn/fw/lib/tracer_node.js
import { Data, Interest, Nack } from "@ndn/packet";
import { console } from "@ndn/util";
import { Forwarder } from "./forwarder_node.js";
/** Print trace logs from Forwarder events. */
export class Tracer {
    static enable(opts = {}) {
        return new Tracer(opts);
    }
    output;
    fw;
    constructor({ output = console, fw = Forwarder.getDefault(), face = true, prefix = true, ann = true, pkt = true, }) {
        this.output = output;
        this.fw = fw;
        if (face) {
            this.fw.addEventListener("faceadd", this.faceadd);
            this.fw.addEventListener("facerm", this.facerm);
        }
        if (prefix) {
            this.fw.addEventListener("prefixadd", this.prefixadd);
            this.fw.addEventListener("prefixrm", this.prefixrm);
        }
        if (ann) {
            this.fw.addEventListener("annadd", this.annadd);
            this.fw.addEventListener("annrm", this.annrm);
        }
        if (pkt) {
            this.fw.addEventListener("pktrx", this.pktrx);
            this.fw.addEventListener("pkttx", this.pkttx);
        }
    }
    disable() {
        this.fw.removeEventListener("faceadd", this.faceadd);
        this.fw.removeEventListener("facerm", this.facerm);
        this.fw.removeEventListener("prefixadd", this.prefixadd);
        this.fw.removeEventListener("prefixrm", this.prefixrm);
        this.fw.removeEventListener("annadd", this.annadd);
        this.fw.removeEventListener("annrm", this.annrm);
        this.fw.removeEventListener("pktrx", this.pktrx);
        this.fw.removeEventListener("pkttx", this.pkttx);
    }
    faceadd = ({ face }) => {
        this.output.log(`+Face ${face}`);
    };
    facerm = ({ face }) => {
        this.output.log(`-Face ${face}`);
    };
    prefixadd = ({ face, prefix }) => {
        this.output.log(`${face} +Prefix ${prefix}`);
    };
    prefixrm = ({ face, prefix }) => {
        this.output.log(`${face} -Prefix ${prefix}`);
    };
    annadd = ({ name }) => {
        this.output.log(`+Announcement ${name}`);
    };
    annrm = ({ name }) => {
        this.output.log(`-Announcement ${name}`);
    };
    pktrx = ({ face, packet }) => {
        this.pkt(face, packet, ">");
    };
    pkttx = ({ face, packet }) => {
        this.pkt(face, packet, "<");
    };
    pkt(face, pkt, dir) {
        switch (true) {
            case pkt.l3 instanceof Interest: {
                const act = pkt.cancel ? "Cancel" :
                    pkt.reject ? `Reject(${pkt.reject})` :
                        "I";
                this.output.log(`${face} ${dir}${act} ${interestToString(pkt.l3)}`);
                break;
            }
            case pkt.l3 instanceof Data: {
                const { name } = pkt.l3;
                this.output.log(`${face} ${dir}D ${name}`);
                break;
            }
            case pkt.l3 instanceof Nack: {
                const { interest, reason } = pkt.l3;
                this.output.log(`${face} ${dir}N ${interestToString(interest)}~${reason}`);
                break;
            }
        }
    }
}
function interestToString({ name, canBePrefix, mustBeFresh }) {
    return `${name}${canBePrefix ? "[P]" : ""}${mustBeFresh ? "[F]" : ""}`;
}
